source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -13,17 +16,49 @@ namespace Envoy {
 namespace Filter {
 namespace Http {
 
+namespace {
+void validateTypeUrlHelper(const std::string& type_url,
+                           const absl::flat_hash_set<std::string> require_type_urls) {
+  if (!require_type_urls.contains(type_url)) {
+    throw EnvoyException(fmt::format("Error: filter config has type URL {} but expect {}.",
+                                     type_url, absl::StrJoin(require_type_urls, ", ")));
+  }
+}
+
+// Updates each provider with update_cb, triggering done_cb once this has been applied to all
+// providers.
+void updateProviders(
+    const absl::flat_hash_set<DynamicFilterConfigProviderImpl*>& filter_config_providers,
+    std::function<void(DynamicFilterConfigProviderImpl*, std::function<void()>)> update_cb,
+    std::function<void()> done_cb) {
+
+  auto remaining_providers =
+      std::make_shared<std::atomic<uint64_t>>(filter_config_providers.size());
+  for (auto* provider : filter_config_providers) {
+    update_cb(provider, [remaining_providers, done_cb] {
+      if (--(*remaining_providers) == 0) {
```

Comment:
Even thought this int is atomic, can't this still race, e.g. imagine we have two threads A and B. They both see 2 remaining. A decrements, then B decrements, then A reads then B reads and they both see 0?

I think you need atomic compare/exchange, but even then, I rather we didn't review this kind of code more than once, so suggest factoring out to something in `source/common/common`, i.e. some sort of "apply to a container's elements in parallel and callback when done"..

Reply:
Moved this to common/common and made it not thread safe since I changed it to be invoked as a completion callback for the runOnAllThreads call, which means it should get invoked on the main thread. I can see an argument for wanting to keep it thread safe just in case, but lmk what you think